rcl_action/src/rcl_action/action_client.c

```diff
@@ -649,6 +685,251 @@ rcl_action_client_wait_set_get_entities_ready(
   return RCL_RET_OK;
 }
 
+#define UUID_FMT \
+  "%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X"
+
+#define UUID_FMT_ARGS(uuid) \
+  uuid[0], \
+  uuid[1], \
+  uuid[2], \
+  uuid[3], \
+  uuid[4], \
+  uuid[5], \
+  uuid[6], \
+  uuid[7], \
+  uuid[8], \
+  uuid[9], \
+  uuid[10], \
+  uuid[11], \
+  uuid[12], \
+  uuid[13], \
+  uuid[14], \
+  uuid[15]
+
+static char*
+to_uuid_string(const uint8_t * uuid) {
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  char * uuid_str = rcutils_format_string(allocator, UUID_FMT, UUID_FMT_ARGS(uuid));
+  return uuid_str;
+}
+
+static
+rcl_ret_t set_content_filtered_topic(
+  const rcl_action_client_t * action_client)
+{
+  rcl_ret_t ret;
+  rcutils_ret_t rcutils_ret;
+  uint8_t * uuid = NULL;
+  char * uuid_str = NULL;
+  size_t size;
+  size_t i;
+  char * feedback_filter = NULL;
+  char * feedback_filter_update = NULL;
+  char * status_filter = NULL;
+  char * status_filter_update = NULL;
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+
+  feedback_filter = rcutils_strdup("", allocator);
+  if (NULL == feedback_filter) {
+    RCL_SET_ERROR_MSG("failed to allocate memory for feedback filter string");
+    ret = RCL_RET_BAD_ALLOC;
+    goto clean;
+  }
+  status_filter = rcutils_strdup("", allocator);
+  if (NULL == status_filter) {
+    RCL_SET_ERROR_MSG("failed to allocate memory for status filter string");
+    ret = RCL_RET_BAD_ALLOC;
+    goto clean;
+  }
+
+  rcutils_ret = rcutils_hash_map_get_size(&action_client->impl->goal_uuids, &size);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    RCL_SET_ERROR_MSG("failed to get map size of goal uuid");
+    ret = RCL_RET_ERROR;
+    goto clean;
+  }
+
+  if (0 == size) {
+    // mainly to reset content filtered topic with empty string
+    goto set_cft;
+  }
+
+  rcutils_ret = rcutils_hash_map_get_next_key_and_data(
+    &action_client->impl->goal_uuids, NULL, &uuid, &uuid_str);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    RCL_SET_ERROR_MSG("failed to get first item from the map of goal uuid");
+    ret = RCL_RET_ERROR;
+    goto clean;
+  }
+
+  feedback_filter_update =
+    rcutils_format_string(allocator, "goal_id.uuid = &hex(%s)", uuid_str);
+  if (NULL == feedback_filter_update) {
+    RCL_SET_ERROR_MSG("failed to fromat string for feedback filter");
+    ret = RCL_RET_BAD_ALLOC;
+    goto clean;
+  }
+  allocator.deallocate(feedback_filter, allocator.state);
+  feedback_filter = feedback_filter_update;
+
+  for (i = 0; i < size; ++i) {
```

Comment:
Does this even work for multiple action clients request the goal?

1. client-A requests goal-1-A.
2. client-B requests goal-2-B.
3. client-A requests goal-3-A.
4. client-B requests goal-4-B.

as far as i see, example on client-A. it will try to do filtering 1 & 2 which are goal-1-A and goal-2-B on the server side list because client-A/B cannot know the array size on server side? so in this case, we would need probably like a wildcard filtering expression?

Reply:
> Does this even work for multiple action clients request the goal?

Yes, it does.

> cannot know the array size on server side?

Sorry? I feel that client-A/B does not need to know the array size on server side.
the feedback and status subscribers are on client side, and they know the array size of goal_id.

- Client A request goal-1-A with UUID-1-A
  - it will set feedback filter with (goal_id.uuid=UUID-1-A) and status filter with (status_list[0].goal_info.goal_id.uuid=UUID-1-A)
- Client B request goal-2-B with UUID-2-B
  - it will set feedback filter with (goal_id.uuid=UUID-2-B) and status filter with (status_list[0].goal_info.goal_id.uuid=UUID-2-B)
- Client A request goal-3-A with UUID-3-A ( if goal-1-A is still processed )
  - it will set feedback filter with (goal_id.uuid=UUID-1-A or goal_id.uuid=UUID-3-A) and status filter with (status_list[0].goal_info.goal_id.uuid=UUID-1-A or status_list[1].goal_info.goal_id.uuid=UUID-3-A or status_list[0].goal_info.goal_id.uuid=UUID-3-A or status_list[1].goal_info.goal_id.uuid=UUID-1-A)
- Client B request goal-4-B with UUID-4-B
  - it will set feedback filter with (goal_id.uuid=UUID-2-B or goal_id.uuid=UUID-4-B) and status filter with (status_list[0].goal_info.goal_id.uuid=UUID-2-B or status_list[1].goal_info.goal_id.uuid=UUID-4-B or status_list[0].goal_info.goal_id.uuid=UUID-4-B or status_list[1].goal_info.goal_id.uuid=UUID-2-B)

What do you think?

>  so in this case, we would need probably like a wildcard filtering expression?

Sorry, I can't find it to support array wildcard filtering expression? So I add the commend for the https://github.com/iuhilnehc-ynos/rclcpp/blob/4ec21c2414c3c8677445a8bcc8cf9b654cbc65e4/rclcpp/src/rclcpp/time_source.cpp#L126-L133

Reply:
After reconsideration, yes, it has the issues there. I'll find a way to support this wildcard filtering expression for array.

Reply:
sorry to be late to get back to this. yeah, the problem is with `index` of `status_list`, semantic of index is different between writer and reader.

Reply:
as far as i can see from https://community.rti.com/static/documentation/connext-dds/5.2.3/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/SQL_Filter_Expression_Notation.htm#topics_3596570972_161086, it does not seem to support the wildcard for array nor sequence.

Reply:
> ask RTI to support this kind of feature. (Actually, the "&hex" for UUID is also an extension for the content filtered topic on rti_connext, other DDS implementation might not support this feature.)

probably 1st step is just ask their opinion is good. with some background about CFT feature, could you create question 
on https://github.com/rticommunity/rmw_connextdds/issues?

> change action message type (goal request) to transfer node name, but it will change the original action design. (One design needs to update another design, is it OK to go for it? If so, do you think we need to update the action design first, and then go back to finish this CFT feature.)

probably we can skip `status` topic at this time, only `feedback`. i would say, CFT is still really good feature for user subscription, parameter events and action feedback topic. so let's finish that 1st. then we can come back to this topic for more additional enhancement with CFT requesting action status message modification. how about that?

Reply:
> could you create question
on https://github.com/rticommunity/rmw_connextdds/issues?

I have opened a feature request on https://github.com/rticommunity/rmw_connextdds/issues/11.

> how about that?

Thanks for your understanding.

Reply:
@iuhilnehc-ynos 

how about we put this `status` topic enhancement as in `T.B.D` but we can improve feedback? we will leave the description as `T.B.D` or `W.I.P` in official PR? how does it sound?